sdk/servicebus/azure-servicebus/azure/servicebus/exceptions.py

```diff
@@ -74,145 +68,116 @@ def _error_handler(error):
-    if error.condition in _NO_RETRY_ERRORS:
-        return errors.ErrorAction(retry=False)
-    return errors.ErrorAction(retry=True)
-
-
-def _handle_amqp_connection_error(logger, exception, handler):
-    # Handle all exception inherited from uamqp.errors.AMQPConnectionError
-    error_need_close_handler = True
-    error_need_raise = False
-    error = None
-    if isinstance(exception, errors.LinkDetach) and exception.condition in _AMQP_SESSION_ERROR_CONDITIONS:
-        # In session lock lost or no active session case, we don't retry, close the handler and raise the error
-        error_need_raise = True
-        if exception.condition == SESSION_LOCK_LOST:
-            try:
-                session_id = handler._session_id  # pylint: disable=protected-access
-            except AttributeError:
-                session_id = None
-            error = SessionLockExpired("Connection detached - lock on Session {} lost.".format(session_id))
-        elif exception.condition == SESSION_LOCK_TIMEOUT:
-            error = NoActiveSession("Queue has no active session to receive from.")
-    elif isinstance(exception, (errors.LinkDetach, errors.ConnectionClose)):
-        # In other link detach and connection case, should retry
-        logger.info("Handler detached due to exception: (%r).", exception)
-        if exception.condition == constants.ErrorCodes.UnauthorizedAccess:
-            error = ServiceBusAuthorizationError(str(exception), error=exception)
-        elif exception.condition == constants.ErrorCodes.NotAllowed and 'requires sessions' in str(exception):
-            message = str(exception) + '\n\nsession_id must be set when getting a receiver for sessionful entity.'
-            error = ServiceBusConnectionError(message, error=exception)
-        else:
-            error = ServiceBusConnectionError(str(exception), error=exception)
-    elif isinstance(exception, errors.MessageHandlerError):
-        logger.info("Handler error: (%r).", exception)
-        error = ServiceBusConnectionError(str(exception), error=exception)
+        return AMQPErrors.ErrorAction(retry=True, backoff=4)
+    if error.condition in _NO_RETRY_CONDITION_ERROR_CODES:
+        return AMQPErrors.ErrorAction(retry=False)
+    return AMQPErrors.ErrorAction(retry=True)
+
+
+def _handle_amqp_exception_with_condition(
+    logger,
+    condition,
+    description,
+    exception=None,
+    status_code=None
+):
+    #
+    # handling AMQP Errors that have the condition field or the mgmt handler
+    logger.info("AMQP error occurred: (%r), condition: (%r), description: (%r).", exception, condition, description)
+    if condition == AMQPErrorCodes.NotFound:
+        # handle NotFound error code
+        error_cls = ServiceBusCommunicationError \
+            if isinstance(exception, AMQPErrors.AMQPConnectionError) else MessagingEntityNotFoundError
+    elif condition == AMQPErrorCodes.ClientError and 'timed out' in str(exception):
```

Comment:
Just checking there's no more reliable way to check the timeout than string comparison?  That's just... more fragile than I'd like if there's any other approach, but if we're stuck we're stuck.  (I see we had this in the past so can feel free to add this as a post-GA issue, can even be up-for-grabs to investigate.)

Reply:
As an aside though, if we're OK with this, it makes me wonder if we could do the same to distinguish the different types of session timeouts that we used to have, if they gave any feedback in that exception body.

Reply:
This is somewhat limited by our uamqp implementation as in uamqp send messages we would wrap any error into ClientError and this could be improved in uamqp.

Reply:
issue created: https://github.com/Azure/azure-uamqp-python/issues/184